lib/Sema/TypeCheckPattern.cpp

```diff
@@ -1125,19 +1124,25 @@ bool TypeChecker::coercePatternToType(Pattern *&P, TypeResolution resolution,
     VarDecl *var = NP->getDecl();
     if (var->isInvalid())
       type = ErrorType::get(Context);
-    var->setType(type);
-    // FIXME: wtf
-    if (type->hasTypeParameter())
-      var->setInterfaceType(type);
-    else
-      var->setInterfaceType(type->mapTypeOutOfContext());
 
-    checkTypeModifyingDeclAttributes(var);
-    if (var->getAttrs().hasAttribute<ReferenceOwnershipAttr>())
-      type = var->getType()->getReferenceStorageReferent();
-    else if (!var->isInvalid())
-      type = var->getType();
+    Type interfaceType = type;
+    if (interfaceType->hasArchetype())
```

Comment:
It still seems like we should know?

Reply:
Patterns are funny because we sometimes check them with contextual types (eg inside an expression) and sometimes with interface types. I'd like to clean this up some day (maybe always using interface types and mapping into context as needed is the cleanest approach).

Reply:
I'm kinda surprised we *ever* use interface types. Okay.